kernel/arch/dreamcast/sound/snd_stream.c

```diff
@@ -503,124 +552,153 @@ void snd_stream_stop(snd_stream_hnd_t hnd) {
+    }
 
-                if(needed_samples & 1)
-                    needed_samples = (needed_samples + 2) & ~1;
-            }
-        }
+    data = stream->get_data(hnd, needed_bytes * stream->channels, &got_bytes);
+    process_filters(hnd, &data, &got_bytes);
 
-        if(data == NULL) {
-            /* Fill the "other" buffer with zeros */
-            spu_memset(streams[hnd].spu_ram_sch[0] + (streams[hnd].last_write_pos * 2), 0, needed_samples * 2);
-            spu_memset(streams[hnd].spu_ram_sch[1] + (streams[hnd].last_write_pos * 2), 0, needed_samples * 2);
-            return -3;
-        }
+    if(got_bytes < needed_bytes * stream->channels) {
+        needed_bytes = got_bytes / stream->channels;
+    }
 
-        if(streams[hnd].stereo) {
-            sep_buffer[1] = sep_buffer[0] + (SND_STREAM_BUFFER_MAX / 8);
-        }
+    /* Round it to max sample size and stereo */
+    if(needed_bytes & 7) {
+        needed_bytes = (needed_bytes + 8) & ~7;
+    }
 
-        if (((uintptr_t)data & 31) && (streams[hnd].type == AICA_SM_16BIT || streams[hnd].stereo == 0)) {
-            sep_data(data, needed_samples * 2, streams[hnd].stereo);
-        }
-        else if(streams[hnd].stereo) {
-            if(streams[hnd].type == AICA_SM_16BIT) {
-                snd_pcm16_split(data, sep_buffer[0], sep_buffer[1], needed_samples * 4);
-            }
-            else if(streams[hnd].type == AICA_SM_8BIT) {
-                snd_pcm8_split(data, sep_buffer[0], sep_buffer[1], needed_samples * 4);
+    needed_samples = bytes_to_samples(hnd, needed_bytes);
+    write_pos = samples_to_bytes(hnd, stream->last_write_pos);
+
+    if(data == NULL) {
+        /* Fill the "other" buffer with zeros */
+        spu_memset(stream->spu_ram_sch[0] + write_pos, 0, needed_bytes);
+        spu_memset(stream->spu_ram_sch[1] + write_pos, 0, needed_bytes);
+        return -3;
+    }
+
+    mutex_lock_timed(&stream_mutex, LOCK_TIMEOUT_MS);
+
+    if(stream->channels == 2) {
+        sep_buffer[1] = sep_buffer[0] + (SND_STREAM_BUFFER_MAX / 8);
+
+        if(streams[hnd].bitsize == 16) {
+            if((uintptr_t)data & 31) {
+                snd_pcm16_split_unaligned(data,
+                    sep_buffer[0], sep_buffer[1], needed_bytes * 2);
             }
             else {
-                snd_adpcm_split(data, sep_buffer[0], sep_buffer[1], needed_samples * 4);
+                snd_pcm16_split((uint32_t *)data,
+                    sep_buffer[0], sep_buffer[1], needed_bytes * 2);
             }
-        } else {
-            first_dma_buf = data;
-            sep_buffer[1] = data;
+        }
+        else if(streams[hnd].bitsize == 8) {
+            snd_pcm8_split(data, sep_buffer[0], sep_buffer[1], needed_bytes * 2);
+        }
+        else if(streams[hnd].bitsize == 4) {
+            snd_adpcm_split(data, sep_buffer[0], sep_buffer[1], needed_bytes * 2);
         }
 
+        dcache_purge_range((uint32)sep_buffer[0], needed_bytes);
```

Comment:
```suggestion
        dcache_purge_range((uint32_t)sep_buffer[0], needed_bytes);
```

Reply:
Do this instead, the functions have been updated to take in uintptr_t instead of uint32_t
```suggestion
        dcache_purge_range((uintptr_t)sep_buffer[0], needed_bytes);
```